src/metabase/models/user_key_value.clj

```diff
@@ -0,0 +1,121 @@
+        (mc/encode ::types/user-key-value
+                   kvp
+                   (mtx/transformer
+                    (mtx/default-value-transformer)
+                    {:name :database}))]
+    (t2/with-transaction [_]
+      (if (t2/select-one :model/UserKeyValue :user_id user-id :namespace namespace :key key)
+        (t2/update! :model/UserKeyValue :user_id user-id :namespace namespace :key key {:value value
+                                                                                        :expires_at expires-at})
+        (try
+          (t2/insert! :model/UserKeyValue {:user_id user-id
+                                           :namespace namespace
+                                           :key key
+                                           :value value
+                                           :expires_at expires-at})
+          ;; in case we caught a duplicate key exception (a row was inserted between our read and write), try updating
+          (catch Exception _
+            (t2/update! :model/UserKeyValue :user_id user-id :namespace namespace :key key {:value value
+                                                                                            :expires_at expires-at})))))
+    value))
+
+(mu/defn delete!
+  "Deletes a KV-pair"
+  [user-id :- :int
+   namespace :- :string
+   k :- :string]
+  (t2/delete! :model/UserKeyValue :namespace namespace :user_id user-id :key k))
+
+(mu/defn retrieve
+  "Retrieves a KV-pair"
+  [user-id :- :int
+   namespace :- :string
+   k :- :string]
+  (when-let [ukv
+             (t2/select-one :model/UserKeyValue
+                            {:where
+                             [:and
+                              [:= :user_id user-id]
+                              [:= :namespace namespace]
+                              [:= :key k]
+                              [:or
+                               [:>= :expires_at :%now]
+                               [:= :expires_at nil]]]})]
+    (:value (mc/decode ::types/user-key-value
+                       ukv
+                       (mtx/transformer
+                        (mtx/default-value-transformer)
+                        {:name :database})))))
+
+(mu/defn retrieve-all
+  "Retrieves all KV-pairs in a namespace"
+  [user-id :- :int
+   namespace :- :string]
+  (when-let [kvs (seq (t2/select :model/UserKeyValue
+                                 {:where
+                                  [:and
+                                   [:= :user_id user-id]
+                                   [:= :namespace namespace]
+                                   [:or
+                                    [:>= :expires_at :%now]
+                                    [:= :expires_at nil]]]}))]
+    (let [parsed-kvs (mc/decode [:sequential ::types/user-key-value]
+                                kvs
+                                (mtx/transformer
+                                 (mtx/default-value-transformer)
```

Comment:
A test with default value would be nice. It would also show people unfamiliar with that functionality how to set those up. Maybe tossing in this link would help with that too? `https://github.com/metosin/malli?tab=readme-ov-file#default-values`